### PR TITLE
New build system

### PR DESCRIPTION
**Pull request by NikitaRM21**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,46 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (ssmolov) [inline CMakeLists.txt]: Here's what we have in our C++ project for that we want to use staccato:
```
cmake_minimum_required(VERSION 3.13)
```
Let's use the same version for CMake here.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+add_compile_options(-O3 -funroll-all-loops)
+
+option(BUILD_SHARED_LIBS "Build shared libraries (ON) or static libraries (OFF)" OFF)
+
+# Check if building library is shared 
+if(BUILD_SHARED_LIBS)
+    if(DEFINED ENV{UTOPIA_HOME})
+        message(STATUS "UTOPIA_HOME is set to: $ENV{UTOPIA_HOME}")
```
Comment (ssmolov) [inline CMakeLists.txt]: There is no need in checking `UTOPIA_HOME` here.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        message(FATAL_ERROR "UTOPIA_HOME environment variable is not set.")
+    endif()
+    
+else()
+    add_library(STACCATO STATIC ${SRCS})
+endif()
+
+target_include_directories(STACCATO
+    PUBLIC
+    #path/to/your/cudd/include/dir
```
Comment (ssmolov) [inline CMakeLists.txt]: What is this commented code for?

File: include/DSD.h
Comment (ssmolov) [inline include/DSD.h:1]: This empty file (and the following) were removed in previous PR. There is no need in such duplication.

File: Makefile [hunk 4]
```diff
@@ -1,62 +0,0 @@
...
-
-CFLAGS   = -c -O3 -funroll-all-loops $(CUDD_INC)
-CC       = g++ 
-
-	
-$(LIB): $(OBJ)
-	ar rv $(LIB) $(OBJ) 		
-
-clean: 
-	rm -rf $(OBJ) $(LIB) sample
```
Comment (ssmolov) [inline Makefile:47]: What is an equivalent for this command at CMake-based system?

File: Makefile [hunk 5]
```diff
@@ -1,62 +0,0 @@
...
-CUDD_LIB = -L/path/to/cudd-2.4.2/cudd
-EPD_LIB = -L/path/to/cudd-2.4.2/epd
-ST_LIB = -L/path/to/cudd-2.4.2/st
-UTIL_LIB = -L/path/to/cudd-2.4.2/util
-MTR_LIB = -L/path/to/cudd-2.4.2/mtr
-DSD_LIB = -L/path/to/STACCATO-1.2/lib
-	
-test:
-	$(CC) src/sample.c $(CUDD_LIB) $(MTR_LIB) $(ST_LIB) $(EPD_LIB) $(UTIL_LIB) $(DSD_LIB) \
-	-lcudd -lmtr -lst -lepd -lutil -ldsd $(CUDD_INC) -o sample
```
Comment (ssmolov) [inline Makefile]: What is an equivalent for this command at CMake-based system?

File: Makefile [hunk 6]
```diff
@@ -1,62 +0,0 @@
...
-#   02110-1301  USA                                                            #
-#                                                                              #
-#   Stephen Plaza <splaza@umich.edu>                                           #
-#   Valeria Bertacco <valeria@umich.edu>                                       #
-#                                                                              #
-#   University of Michigan                                                     #
-#   Electrical Engineering and Computer Science Dept.                          #
-#   2260 Hayward St.                                                           #
-#   Ann Arbor, MI 48109-2121                                                   #
-################################################################################
```
Comment (ssmolov) [inline Makefile]: Please put this to `CMakeLists.txt` as heading comment

File: INSTALL [hunk 7]
```diff
@@ -30,12 +30,25 @@
...
+CUDD_PATH=example/path/to/cudd.
+
+Example for building shared library:
+$> make BUILD_TYPE=shared CUDD_PATH=example/path/to/cudd
+
+Example for building static library:
+$> make CUDD_PATH=example/path/to/cudd
+
+You can test static library with:
+$> make test CUDD_PATH=example/path/to/cudd
```
Comment (ssmolov) [inline INSTALL:46]: What is `CUDD_PATH`? Is it a path to sources or to *.so?

What information about CUDD was used at previous (non-modified) STACCATO?

File: Makefile [hunk 8]
```diff
@@ -29,34 +29,57 @@
...
+else
+    LIB = libSTACCATO.so
+endif
 
+SRC = $(shell ls src/*.c)
+OBJ = $(SRC:%.c=%.o)
 
-CFLAGS   = -c -O3 -funroll-all-loops $(CUDD_INC)
-CC       = g++ 
+CUDD_INC = $(CUDD_PATH)/include
```
Comment (ssmolov) [inline Makefile:44]: CUDD_INC -> CUDD_INCLUDE

File: Makefile [hunk 9]
```diff
@@ -29,34 +29,57 @@
...
 
-	
+ifeq ($(BUILD_TYPE), static)
+	CFLAGS	= -c -O3 -funroll-all-loops -I$(CUDD_INC)
```
Comment (ssmolov) [inline Makefile:47]: Modified indent, see previous `ifeq` block.

File: Makefile [hunk 10]
```diff
@@ -29,34 +29,57 @@
...
+else
+	CFLAGS	= -c -O3 -funroll-all-loops -fPIC -I$(CUDD_INC)
```
Comment (ssmolov) [inline Makefile:49]: Modified indent, see previous `ifeq` block.

File: INSTALL (hunk 7 above)
Comment (NikitaRM21) [reply INSTALL:46]: CUDD_PATH it is a path to directory with CUDD sources. 
In non-modified STACCATO they used the same information: path to CUDD headers (cudd/include directory).

Comment (ssmolov) [reply INSTALL:46]: Ok. It should be concretized here what "CUDD directory" is to avoid misunderstanding.

**Pull request closed without merging**